src/Exception/BatchException.php

```diff
@@ -2,61 +2,49 @@
 
 namespace Http\Client\Exception;
 
+use Http\Client\BatchResult;
+
 /**
  * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
  */
-final class BatchException extends TransferException
+final class BatchException extends RuntimeException
 {
     /**
-     * @var TransferException[]
+     * @var BatchResult
      */
-    private $exceptions;
+    private $result;
 
     /**
-     * @param TransferException[] $exceptions
+     * @param BatchResult $result
      */
-    public function __construct(array $exceptions = [])
+    public function __construct(BatchResult $result)
     {
-        parent::__construct('An error occurred when sending multiple requests.');
-
-        foreach ($exceptions as $e) {
-            if (!$e instanceof TransferException) {
-                throw new InvalidArgumentException('Exception is not an instanceof Http\Client\Exception\TransferException');
-            }
-        }
-
-        $this->exceptions = $exceptions;
+        $this->result = $result;
     }
 
     /**
-     * Returns all exceptions
+     * Returns successful responses or null
      *
-     * @return TransferException[]
+     * @return BatchResult
      */
-    public function getExceptions()
+    public function getResult()
     {
-        return $this->exceptions;
+        return $this->result;
     }
 
     /**
-     * Checks if a specific exception exists
+     * Decides whether the result has any failures
      *
-     * @param TransferException $exception
+     * @param BatchResult $result
      *
-     * @return boolean TRUE if there is the exception else FALSE.
+     * @return BatchResult|BatchException
      */
-    public function hasException(TransferException $exception)
+    public static function decideReturnValue(BatchResult $result)
```

Comment:
> and it would be even more weird to return a BatchException from the client. 

You are right, I wasn't thinking when I wrote this.

I agree with the rest: if the result has exceptions, throw it in a batch exception.

It is also easier, because you have the same internal logic in sendRequest: ALWAYS collect responses and exceptions in BatchResult and it's not like: if we have an exception, we create a batchexception and put it there and in the end if we have a batchexception, we throw it.